File: impl/go/internal/engine/abilities_test.go
package engine

import (
	"testing"

	"github.com/zaibatsu/zaibatsu-go/internal/domain"
)

// searchGame returns a fresh game with speedrunner-red (has Search) on the core
// and a deterministic block pile.
func searchGame(t *testing.T, pile ...string) (*domain.GameState, *domain.GameData) {
	t.Helper()
	gd := loadOrSkip(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 1})
	s.Cybernet.Pawns = []*domain.PawnOnBoard{}
	s.Cybernet.PlacePawn(&domain.PawnOnBoard{PawnID: "speedrunner-red", OwnerID: "p1", Coord: domain.Coord{Q: 0, R: 0}, SpaceID: "core"})
	s.BlockPile = append([]string{}, pile...)
	return s, gd
}

func TestSearchPlacesTopBlock(t *testing.T) {
	s, gd := searchGame(t, "data-haven")
	origin := domain.Coord{Q: 0, R: 0}
	dir := 2
	rot := rotFacing(t, gd, "data-haven", dir)
	pb, err := Search(s, gd, "speedrunner-red", dir, rot)
	if err != nil {
		t.Fatalf("Search: %v", err)
	}
	if pb.BlockID != "data-haven" {
		t.Errorf("placed %q, want data-haven", pb.BlockID)
	}
	if pb.Coord != origin.Neighbor(dir) {
		t.Errorf("placed at %v, want %v", pb.Coord, origin.Neighbor(dir))
	}
	if len(s.BlockPile) != 0 {
		t.Errorf("pile should be empty after search, got %d", len(s.BlockPile))
	}
	if s.Cybernet.At(origin.Neighbor(dir)) == nil {
		t.Error("cell not occupied after search")
	}
}

func TestSearchInvalidPlacementDoesNotConsume(t *testing.T) {
	s, gd := searchGame(t, "data-haven")
	// data-haven edges [T,F,T,T,F,T]; find a rotation whose edge facing back on
	// dir 0 is FALSE so placement is rejected.
	b, _ := gd.BlockByID("data-haven")
	badRot := -1
	for rot := 0; rot < 6; rot++ {
		if !EdgeHasSpace(b, rot, domain.Opposite(0)) {
			badRot = rot
			break
		}
	}
	if badRot < 0 {
		t.Skip("no unconnected rotation")
	}
	if _, err := Search(s, gd, "speedrunner-red", 0, badRot); err == nil {
		t.Error("expected rejection for unconnected orientation")
	}
	if len(s.BlockPile) != 1 {
		t.Errorf("failed search must not consume the block; pile = %d", len(s.BlockPile))
	}
}

func TestSearchEmptyPile(t *testing.T) {
	s, gd := searchGame(t) // empty pile
	if _, err := Search(s, gd, "speedrunner-red", 0, 0); err == nil {
		t.Error("expected error searching with an empty pile")
	}
}

func TestSearchRequiresSearchAbility(t *testing.T) {
	s, gd := searchGame(t, "data-haven")
	// drone-turret has no Search ability.
	s.Cybernet.Pawns = []*domain.PawnOnBoard{}
	s.Cybernet.PlacePawn(&domain.PawnOnBoard{PawnID: "drone-turret", OwnerID: "p1", Coord: domain.Coord{Q: 0, R: 0}, SpaceID: "core"})
	if _, err := Search(s, gd, "drone-turret", 0, 0); err == nil {
		t.Error("expected error: drone-turret cannot Search")
	}
}

func TestValidSearchPlacements(t *testing.T) {
	s, gd := searchGame(t, "data-haven")
	opts, err := ValidSearchPlacements(s, gd, "speedrunner-red")
	if err != nil {
		t.Fatalf("ValidSearchPlacements: %v", err)
	}
	if len(opts) == 0 {
		t.Fatal("expected valid placements around the core")
	}
	for _, o := range opts {
		if err := CanPlace(s.Cybernet, gd, domain.Coord{Q: 0, R: 0}, o.Dir, "data-haven", o.Rotation); err != nil {
			t.Errorf("returned invalid option %+v: %v", o, err)
		}
	}
}

func TestRebootReturnsPawnToCore(t *testing.T) {
	gd := loadOrSkip(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 1})
	origin := domain.Coord{Q: 0, R: 0}
	// speedrunner-red has the Reboot ability; mark it eliminated.
	s.Cybernet.RemovePawn("speedrunner-red") // no-op if absent
	s.Eliminated = []string{"speedrunner-red"}
	pob, err := Reboot(s, gd, "speedrunner-red", "p1")
	if err != nil {
		t.Fatalf("Reboot: %v", err)
	}
	if pob.Coord != origin {
		t.Errorf("rebooted at %v, want Central Core %v", pob.Coord, origin)
	}
	if pob.OwnerID != "p1" {
		t.Errorf("rebooted owner %q, want p1", pob.OwnerID)
	}
	if s.Cybernet.PawnByID("speedrunner-red") == nil {
		t.Error("rebooted pawn should be on the board")
	}
	if len(s.Eliminated) != 0 {
		t.Errorf("Eliminated pool should be empty, got %v", s.Eliminated)
	}
}

func TestRebootRejectsNonEliminated(t *testing.T) {
	gd := loadOrSkip(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 1})
	if _, err := Reboot(s, gd, "speedrunner-red", "p1"); err == nil {
		t.Error("expected error rebooting a pawn that is not eliminated")
	}
}

func TestRebootRequiresRebootAbility(t *testing.T) {
	gd := loadOrSkip(t)
	s, _ := NewGame(Config{Data: gd, PlayerNames: []string{"A", "B"}, Seed: 1})
	// drone-turret has no Reboot ability.
	s.Eliminated = []string{"drone-turret"}
	if _, err := Reboot(s, gd, "drone-turret", "p1"); err == nil {
		t.Error("expected error: drone-turret cannot Reboot")
	}
}

File: impl/ts/test/abilities.test.ts
import { describe, expect, test } from "bun:test";
import { loadDefault } from "../src/data/index.ts";
import { blockById, type GameData, type GameState } from "../src/domain/types.ts";
import { neighbor, opposite, type Coord } from "../src/domain/hex.ts";
import {
  canPlace,
  newGame,
  reboot,
  search,
  validSearchPlacements,
} from "../src/engine/index.ts";

const data: GameData = loadDefault("speedrunners");
const ORIGIN: Coord = { q: 0, r: 0 };

function rotFacing(blockId: string, dir: number): number {
  const b = blockById(data, blockId)!;
  for (let rot = 0; rot < 6; rot++) {
    const local = (((opposite(dir) - rot) % 6) + 6) % 6;
    if (b.edges![local]) return rot;
  }
  throw new Error(`block ${blockId} cannot face direction ${dir}`);
}

function searchGame(...pile: string[]): GameState {
  const s = newGame({ data, playerNames: ["A", "B"], seed: 1 });
  s.cybernet.pawns = [];
  s.cybernet.placePawn({ pawnId: "speedrunner-red", ownerId: "p1", coord: { ...ORIGIN }, spaceId: "core" });
  s.blockPile = [...pile];
  return s;
}

describe("search", () => {
  test("places the top block of the pile", () => {
    const s = searchGame("data-haven");
    const dir = 2;
    const pb = search(s, data, "speedrunner-red", dir, rotFacing("data-haven", dir));
    expect(pb.blockId).toBe("data-haven");
    expect(pb.coord).toEqual(neighbor(ORIGIN, dir));
    expect(s.blockPile.length).toBe(0);
    expect(s.cybernet.at(neighbor(ORIGIN, dir))).toBeDefined();
  });

  test("an invalid placement does not consume the block", () => {
    const s = searchGame("data-haven");
    const b = blockById(data, "data-haven")!;
    let badRot = -1;
    for (let rot = 0; rot < 6; rot++) {
      const local = (((opposite(0) - rot) % 6) + 6) % 6;
      if (!b.edges![local]) {
        badRot = rot;
        break;
      }
    }
    expect(badRot).toBeGreaterThanOrEqual(0);
    expect(() => search(s, data, "speedrunner-red", 0, badRot)).toThrow();
    expect(s.blockPile.length).toBe(1);
  });

  test("empty pile throws", () => {
    const s = searchGame();
    expect(() => search(s, data, "speedrunner-red", 0, 0)).toThrow();
  });

  test("requires the Search ability", () => {
    const s = searchGame("data-haven");
    s.cybernet.pawns = [];
    s.cybernet.placePawn({ pawnId: "drone-turret", ownerId: "p1", coord: { ...ORIGIN }, spaceId: "core" });
    expect(() => search(s, data, "drone-turret", 0, 0)).toThrow();
  });

  test("validSearchPlacements returns only valid options", () => {
    const s = searchGame("data-haven");
    const opts = validSearchPlacements(s, data, "speedrunner-red");
    expect(opts.length).toBeGreaterThan(0);
    for (const o of opts) {
      expect(canPlace(s.cybernet, data, ORIGIN, o.dir, "data-haven", o.rotation)).toBeUndefined();
    }
  });
});

describe("reboot", () => {
  test("returns an eliminated pawn to the Central Core", () => {
    const s = newGame({ data, playerNames: ["A", "B"], seed: 1 });
    s.cybernet.removePawn("speedrunner-red");
    s.eliminated = ["speedrunner-red"];
    const pob = reboot(s, data, "speedrunner-red", "p1");
    expect(pob.coord).toEqual(ORIGIN);
    expect(pob.ownerId).toBe("p1");
    expect(s.cybernet.pawnById("speedrunner-red")).toBeDefined();
    expect(s.eliminated.length).toBe(0);
  });

  test("rejects a non-eliminated pawn", () => {
    const s = newGame({ data, playerNames: ["A", "B"], seed: 1 });
    expect(() => reboot(s, data, "speedrunner-red", "p1")).toThrow();
  });

  test("requires the Reboot ability", () => {
    const s = newGame({ data, playerNames: ["A", "B"], seed: 1 });
    s.eliminated = ["drone-turret"];
    expect(() => reboot(s, data, "drone-turret", "p1")).toThrow();
  });
});

describe("ability removal via attachment", () => {
  test("an attachment that removes Search blocks it", () => {
    const d = loadDefault("speedrunners");
    const s = newGame({ data: d, playerNames: ["A", "B"], seed: 1 });
    s.cybernet.pawns = [];
    s.cybernet.placePawn({ pawnId: "speedrunner-red", ownerId: "p1", coord: { ...ORIGIN }, spaceId: "core" });
    s.blockPile = ["data-haven"];
    d.cards[0].attach = { as: "pawn", slot: "add-on", removes: ["search"] };
    s.cybernet.pawnById("speedrunner-red")!.attachments = [{ cardId: d.cards[0].id, slot: "add-on", bonusPaid: 0 }];
    expect(() => search(s, d, "speedrunner-red", 2, rotFacing("data-haven", 2))).toThrow();
  });
});
